### DISCUSSION TITLE
Let newly created emacsclient window auto focused

### DISCUSSION THREAD

**Discussion by robotechdy** [d12frosted/homebrew-emacs-plus#636] [Q&A]
Firstly, thanks for providing the emacs-plus. I installed it on my Intel Mac, M1 Mac. They all work very well. I can start the emacs service with `emacs --daemon` or `brew services start emacs-plus@29`. Then, I can quickly open any number of client windows with `emacsclient -c`. Every emacs client window is GUI enabled -- nice!

There is one minor issue that bothers me: when I use `emacsclient -c` to open a new client window, the emacs client window is not  auto-focused. I need to manually focus on the newly created window (with a keyboard shortcut or mouse click). I tried to start `emacsclient -c` with Raycast or as a shortcut in `/Applications`, still cannot make the newly created emacsclient window auto-focused.

**Comment by robotechdy** [comment]
@d12frosted  Any suggestion? :)

**Reply by d12frosted** [reply]
Unfortunately, no immediate ideas. I need to investigate, but I am short on free time 🥲

A few seemingly relevant questions:

- does it happen in a single space, or do you use multiple spaces? Just curious if the new window (in macos terms) is focused when you open it the same space as other emacs windows. 
- same about monitors

I have some directions:

- check macos system settings with regards to spaces or new windows. I am in the car atm, so can't check the proper name
- check if yabai can help with focusing new emacs windows;

Sorry, not very helpful.

P. S. Please keep us posted 🙏

**Reply by robotechdy** [reply]
@d12frosted for "space", do you mean virtual desktop? I have multiple virtual desktops. 
For monitors, I only turn on one monitor.

**Comment by robotechdy** [comment]
I installed and configured [hammerspoon](https://www.hammerspoon.org/) to implement an auto-focus feature. Following the steps below, anyone can have this feature.

1. Install hammerspoon with `brew install hammerspoon`. Then, make sure your Mac allow it running in background.
2. Open `.hammerspoon/init.lua`, and copy the code below into it.
```
-- Importing necessary modules from Hammerspoon
local application = require 'hs.application'
local window = require 'hs.window'

-- Setting up a filter for Emacs windows
emacsFilter = hs.window.filter.new('Emacs')

-- Function to focus on the newest Emacs window
function focusNewEmacsWindow()
    -- Getting all Emacs windows, sorted by the time they were created (newest first)
    local windows = emacsFilter:getWindows(hs.window.filter.sortByCreatedLast)
    -- If there are any Emacs windows, focus on the newest one
    if #windows > 0 then
        windows[1]:focus()
    end
end

-- Subscribing to window creation events for Emacs windows
-- Calls focusNewEmacsWindow function each time a new Emacs window is created
emacsFilter:subscribe(hs.window.filter.windowCreated, focusNewEmacsWindow)
```
3. Restart the hammerspoon. Then, whenever you create a new emacs client with `emacsclient -c`, it is automatically focused.

Here above is a bypass solution.
@d12frosted Wish we could have a native built-in solution in the future.  :)
